### PR TITLE
Linked list refactor

### PR DESCRIPTION
**Pull request by weapdiv-david**
### What is the context of this PR?

PR to unlbock @warrenbailey 

This PR refactors the linked list handling to ensure that state is only modified on POST, and that the state can be traversed out-of-order.  It lays the foundation of the summary improvements story.

It also fixes the bug whereby POSTing to the summary page, transferred the user to the thank you page.
### How to review

1) Check out the branch
2) Run the tests
- [ ] Do all commits have sensible commit messages?
- [ ] Is all new code unit tested?
- [ ] Are there integration tests if needed?
- [ ] Is there sufficient logging?
- [ ] Is there documentation or is the code self-describing?
### Who worked on the PR

@weapdiv-david

### CODE REVIEW

File: app/questionnaire/questionnaire_manager.py [hunk 1]
```diff
@@ -95,6 +105,8 @@ def go_to_state(self, item_id):
             self._create_new_state(item_id)
         StateManager.save_state(self)
 
+        StateManager.save_state(self)
```
Comment (collisdigital) [inline app/questionnaire/questionnaire_manager.py]: Duplicate save line 106

Comment (weapdiv-david) [reply app/questionnaire/questionnaire_manager.py]: This is fixed in a later commit... These were cherry picked from the eq-525 branch to unblock warren, so I missed it.  Will fix now

Comment (collisdigital) [reply app/questionnaire/questionnaire_manager.py]: 👍

File: app/questionnaire/questionnaire_manager.py [hunk 2]
```diff
@@ -83,9 +92,10 @@ def go_to_state(self, item_id):
         node = self.get_state(item_id)
         logger.debug("go to state %s", item_id)
         if node:
+            self._current = node
             logger.debug("truncating to node %s", item_id)
```
Comment (iwootten) [inline app/questionnaire/questionnaire_manager.py]: Is this log message still valid?

File: app/questionnaire/questionnaire_manager.py [hunk 3]
```diff
@@ -83,9 +92,10 @@ def go_to_state(self, item_id):
...
-            if node.next:
-                self._truncate(node.next)
+            # if node.next:
```
Comment (iwootten) [inline app/questionnaire/questionnaire_manager.py]: Is there any reason why the commented code can't be removed?

File: app/questionnaire/questionnaire_manager.py [hunk 4]
```diff
@@ -121,30 +131,36 @@ def update_state(self, item_id, user_input):
...
+            if node.next:
+                self._truncate(node.next)
+
+            StateManager.save_state(self)
         else:
             raise TypeError("Can only handle blocks")
 
     def _append(self, node):
         if not self._first:
             self._first = node
```
Comment (iwootten) [inline app/questionnaire/questionnaire_manager.py:146]: Can't the self._current/self._tail assignment be moved below this if?

File: app/questionnaire/questionnaire_manager.py (hunk 2 above)
Comment (weapdiv-david) [reply app/questionnaire/questionnaire_manager.py]: Removed in a later commit, but missed from this when the commits were cherry-picked. Will fix now

File: app/questionnaire/questionnaire_manager.py (hunk 3 above)
Comment (weapdiv-david) [reply app/questionnaire/questionnaire_manager.py]: Removed in a later commit, but missed from this when the commits were cherry-picked. Will fix now

**Pull request merged**